apps/console/src/features/roles/components/edit-role/edit-role-permission-updated.tsx

```diff
@@ -0,0 +1,459 @@
+        isLoading: isSelectedAPIResourceFetchRequestLoading,
+        isValidating: isSelectedAPIResourceFetchRequestValidating,
+        error: selectedAPIResourceFetchRequestError
+    } = useAPIResourceDetails(selectedAPIResourceId);
+
+    useEffect(() => {
+        getExistingAPIResources();
+    }, [ roleObject ]);
+
+    /**
+     * Show error if the API resource fetch request failed.
+     */ 
+    useEffect(() => {
+        if ( selectedAPIResourceFetchRequestError ||  apiResourcsListFetchRequestError) {
+            dispatch(
+                addAlert({
+                    description: t("console:manage.features.roles.addRoleWizard.forms.rolePermission." +
+                        "notifications.fetchAPIResourceError.error.description"),
+                    level: AlertLevels.ERROR,
+                    message: t("console:manage.features.roles.addRoleWizard.forms.rolePermission." +
+                        "notifications.fetchAPIResourceError.error.message")
+                })
+            );
+        }
+    }, [ selectedAPIResourceFetchRequestError, apiResourcsListFetchRequestError ]);
+
+    useEffect(() => {
+        const options: DropdownProps[] = [];
+
+        apiResourcesList?.apiResources?.map((apiResource: APIResourceInterface) => {
+            if (!selectedAPIResources.find((selectedAPIResource: APIResourceInterface) => 
+                selectedAPIResource?.id === apiResource?.id)) {
+                options.push({
+                    key: apiResource.id,
+                    text: apiResource.name,
+                    value: apiResource.id
+                });
+            }
+        });
+
+        setAPIResourcesListOptions(options);
+    }, [ apiResourcesList, selectedAPIResources ]);
+
+    /**
+     * Add API resource to the selected API resources list.
+     */
+    useEffect(() => {
+        if (selectedAPIResource 
+            && !isSelectedAPIResourceFetchRequestLoading 
+            && !isSelectedAPIResourceFetchRequestValidating) {
+            
+            if (!selectedAPIResources.find(
+                (apiResource: APIResourceInterface) => selectedAPIResource?.id === apiResource?.id)) {
+                setSelectedAPIResources([ selectedAPIResource, ...selectedAPIResources ]);
+            }
+
+            selectedAPIResourceId && setSelectedAPIResourceId(undefined);
+        }
+    }, [ selectedAPIResource ]);
+
+    useEffect(() => {
+        getAPIResourceById(initialAPIResourceIds);
+    }, [ initialAPIResourceIds ]);
+        
+    const getAPIResourceById = async (initialAPIResourceIds: string[]): Promise<void> => {
+        const apiResourceIds: string[] = initialAPIResourceIds;
+        const apiResources: APIResourceInterface[] = [];
+
+        if (apiResourceIds.length === 0) {
+            return;
+        }
+
+        for (const apiResourceId of apiResourceIds) {
+            try {
+                apiResources.push(await getAPIResourceDetails(apiResourceId));
```

Comment:
WDYT about utilizing Promise.all to perform `getAPIResourceDetails` concurrently?

Reply:
Will fix this with a follow up PR using SWR. There is an optimization to be done